### PR TITLE
Abstract transactions in tx pool and add staking transaction to pool with error report

### PR DESCRIPTION
**Pull request by Daniel-VDM**
## Issue

The transaction pool only accepts plain transactions and needs to be generalized to handle staking transactions. Moreover, only the leader can report failed staking transaction errors ( similar to #2160 ). Lastly, staking messages are only verified when they are applied, this should be separated so staking msg verification can be done at an earlier stage like the tx pool.

### Highlights
* Abstracts transactions in the tx pool by using a `PoolTransaction` interface. This also required some rework of the `tx_list.go` and `tx_journal.go`. This is most of the PR. 
* Implement methods for `StakingTransactions` to satisfy the tx pool interface. 
* Refactor error reporting in tx pool to handle multiple error sinks.
* Add staking transaction checks in the tx pool and report any errors to the error sink.
* Update consumers and providers of tx pool to interface staking transactions.
* Fix RPC `GetTransactionByHash` to never look in tx pool.
* Add pending staking transaction to hmy API. 
* Remove case when old transactions in tx pool report to error sink even though it is finalized. 
* Refactor staking message verification and staking msg application in `state_transition.go`.
* Make all nodes broadcast transactions if they received a message via RPC.

### Unit Test Coverage

Before:

```
PASS
coverage: 20.3% of statements
ok  	github.com/harmony-one/harmony/core	6.815s
```

```
PASS
coverage: 14.3% of statements
ok  	github.com/harmony-one/harmony/node	1.949s
```

After:

```
PASS
coverage: 20.5% of statements
ok  	github.com/harmony-one/harmony/core	8.504s
```

```
PASS
coverage: 14.7% of statements
ok  	github.com/harmony-one/harmony/node	3.389s
```

### Test/Run Logs

* [Transaction Generation](https://gist.github.com/Daniel-VDM/490c711c982209fd3d654b0ea90bda35) (~500 txs in 60 sec on localnet, all succeeded)
* [Basic Staking Tx test](https://gist.github.com/Daniel-VDM/6a275e3c06147438065282a1ed00f7a6) (create validator, edit validator, delegate, undelegate). This test was done with `GenesisGasLimit uint64 = 10712388` since currently create validator costs more than block gas limit on localnet.

## Operational Checklist

1. **Does this PR introduce backward-incompatible changes to the on-disk data structure and/or the over-the-wire protocol?**. (If no, skip to question 8.)

    **NO**

8. **Does this PR introduce backward-incompatible changes *NOT* related to on-disk data structure and/or over-the-wire protocol?** (If no, continue to question 11.)

    **NO**

11. **Does this PR introduce significant changes to the operational requirements of the node software, such as >20% increase in CPU, memory, and/or disk usage?**

    **NO**

### CODE REVIEW

File: cmd/harmony/main.go [hunk 1]
```diff
@@ -425,7 +425,7 @@ func setupConsensusAndNode(nodeConfig *nodeconfig.ConfigType) *node.Node {
 	return currentNode
 }
 
-func setupBlacklist() (*map[ethCommon.Address]struct{}, error) {
+func setupBlacklist() (map[ethCommon.Address]struct{}, error) {
```
Comment (LeoHChen) [inline cmd/harmony/main.go]: why you have a blacklist change in this PR? Please use a separate one for blacklist fix.

File: core/tx_journal.go [hunk 2]
```diff
@@ -76,33 +100,57 @@ func (journal *txJournal) load(add func([]*types.Transaction) []error) error {
...
+		var tx types.PoolTransaction
+		switch txType, err := stream.Uint(); txType {
+		case plainTxID:
+			tx = new(types.Transaction)
+		case stakingTxID:
+			tx = new(staking.StakingTransaction)
+		default:
+			if err != nil {
+				if err == io.EOF {
+					err = nil
```
Comment (LeoHChen) [inline core/tx_journal.go]: why ignore the EOF error? please comment.

File: core/tx_journal.go [hunk 3]
```diff
@@ -76,33 +100,57 @@ func (journal *txJournal) load(add func([]*types.Transaction) []error) error {
...
+					loadBatch(batch)
+				}
+				return err
+			}
+		}
+
 		if err = stream.Decode(tx); err != nil {
-			if err != io.EOF {
-				failure = err
+			if err == io.EOF {
```
Comment (LeoHChen) [inline core/tx_journal.go]: why ignore EOF?

File: core/tx_list.go [hunk 4]
```diff
@@ -380,11 +380,11 @@ func (h priceHeap) Less(i, j int) bool {
 		return false
 	}
 	// If the prices match, stabilize via nonces (high nonce is worse)
-	return h[i].Nonce() > h[j].Nonce()
+	return h[i].Nonce() > h[i].Nonce()
```
Comment (LeoHChen) [inline core/tx_list.go]: why this change? is it related?

File: core/types/tx_pool.go [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+	staking "github.com/harmony-one/harmony/staking/types"
+)
+
+var (
+	// ErrUnknownPoolTxType is returned when attempting to assert a PoolTransaction to its concrete type
+	ErrUnknownPoolTxType = errors.New("unknown transaction type in tx-pool")
+)
+
+// PoolTransaction is the general transaction interface used by the tx pool
+type PoolTransaction interface {
```
Comment (LeoHChen) [inline core/types/tx_pool.go]: do you need to document each field in this interface?

Comment (sophoah) [reply core/types/tx_pool.go]: Actually comments are good for external parties that do not understanding entirely the code.

File: core/tx_list.go (hunk 4 above)
Comment (Daniel-VDM) [reply core/tx_list.go]: Ah was a typo, thx for catching

File: core/tx_journal.go (hunk 2 above)
Comment (Daniel-VDM) [reply core/tx_journal.go]: ignore EOF to exit reading from journal final without error, will document

File: core/tx_journal.go (hunk 3 above)
Comment (Daniel-VDM) [reply core/tx_journal.go]: mm yes, we should not ignore EOF here, will change

File: cmd/harmony/main.go (hunk 1 above)
Comment (Daniel-VDM) [reply cmd/harmony/main.go]: okay, will remove blacklist change and add in followup pr

File: core/types/tx_pool.go (hunk 5 above)
Comment (Daniel-VDM) [reply core/types/tx_pool.go]: I believe we force all exposed components to have documentation (via golint in travis check?)

File: core/tx_pool.go [hunk 6]
```diff
@@ -675,14 +718,282 @@ func (pool *TxPool) validateTx(tx *types.Transaction, local bool) error {
...
+	}
+}
+
+func (pool *TxPool) validateCreateValidatorMsg(stkMsg *staking.CreateValidator) error {
+	if pool.currentState.IsValidator(stkMsg.ValidatorAddress) {
+		return errValidatorExist
+	}
+
+	// create temp validator to do sanity check
+	validatorWrapper, err := staking.CreateValidatorFromNewMsg(stkMsg, pool.chain.CurrentBlock().Number())
```
Comment (rlan35) [inline core/tx_pool.go]: This few steps is a copy of the actual applyStakingTransaction, better to refactor them into a single function. Otherwise, there can be consistency issue in the future.

File: core/tx_pool.go [hunk 7]
```diff
@@ -675,14 +718,282 @@ func (pool *TxPool) validateTx(tx *types.Transaction, local bool) error {
...
+
+	return w.SanityCheck()
+}
+
+func (pool *TxPool) validateEditValidatorMsg(stkMsg *staking.EditValidator) error {
+	if !pool.currentState.IsValidator(stkMsg.ValidatorAddress) {
+		return errValidatorNotExist
+	}
+
+	validatorWrapper := pool.currentState.GetStakingInfo(stkMsg.ValidatorAddress)
```
Comment (rlan35) [inline core/tx_pool.go]: same here. Should refactor all these logic into a single function that can be called directly.

File: core/tx_pool.go [hunk 8]
```diff
@@ -675,14 +718,282 @@ func (pool *TxPool) validateTx(tx *types.Transaction, local bool) error {
...
+func (pool *TxPool) validateDelegateMsg(stkMsg *staking.Delegate) error {
+	if stkMsg.Amount.Sign() == -1 {
+		return errNegativeAmount
+	}
+
+	if !pool.currentState.IsValidator(stkMsg.ValidatorAddress) {
+		return errValidatorNotExist
+	}
+
+	validatorWrapper := pool.currentState.GetStakingInfo(stkMsg.ValidatorAddress)
```
Comment (rlan35) [inline core/tx_pool.go]: same here.

File: core/tx_pool.go [hunk 9]
```diff
@@ -675,14 +718,282 @@ func (pool *TxPool) validateTx(tx *types.Transaction, local bool) error {
...
+func (pool *TxPool) validateUndelegateMsg(stkMsg *staking.Undelegate) error {
+	if stkMsg.Amount.Sign() == -1 {
+		return errNegativeAmount
+	}
+
+	if !pool.currentState.IsValidator(stkMsg.ValidatorAddress) {
+		return errValidatorNotExist
+	}
+
+	validatorWrapper := pool.currentState.GetStakingInfo(stkMsg.ValidatorAddress)
```
Comment (rlan35) [inline core/tx_pool.go]: same

File: core/tx_pool.go [hunk 10]
```diff
@@ -675,14 +718,282 @@ func (pool *TxPool) validateTx(tx *types.Transaction, local bool) error {
...
+	return nil
+}
+
+func (pool *TxPool) validateCollectRewardsMsg(stkMsg *staking.CollectRewards) error {
+	chain, ok := pool.chain.(*BlockChain)
+	if !ok {
+		return nil // for testing, chain could be testing blockchain
+	}
+
+	delegations, err := chain.ReadDelegationsByDelegator(stkMsg.DelegatorAddress)
```
Comment (rlan35) [inline core/tx_pool.go]: same

File: core/tx_pool.go [hunk 11]
```diff
@@ -1057,10 +1381,6 @@ func (pool *TxPool) promoteExecutables(accounts []common.Address) {
 		for _, tx := range list.Forward(nonce) {
 			hash := tx.Hash()
 			logger.Warn().Str("hash", hash.Hex()).Msg("Removed old queued transaction")
-			if pool.chain.CurrentBlock().Transaction(hash) == nil {
```
Comment (rlan35) [inline core/tx_pool.go]: why are these checks removed?

File: core/tx_pool.go [hunk 12]
```diff
@@ -1242,10 +1557,6 @@ func (pool *TxPool) demoteUnexecutables() {
 		for _, tx := range list.Forward(nonce) {
 			hash := tx.Hash()
 			logger.Warn().Str("hash", hash.Hex()).Msg("Removed old pending transaction")
-			if pool.chain.CurrentBlock().Transaction(hash) == nil {
```
Comment (rlan35) [inline core/tx_pool.go]: same here?

File: core/tx_pool.go (hunk 11 above)
Comment (Daniel-VDM) [reply core/tx_pool.go]: Error report not needed here since error would be caught when tx is added to tx-pool in the first place (from the `validateTx` fn). Also looking up a tx is a linear time op in the number of txs in the block.

File: core/tx_pool.go (hunk 12 above)
Comment (Daniel-VDM) [reply core/tx_pool.go]: addressed in other msg

File: node/node.go [hunk 13]
```diff
@@ -282,50 +279,48 @@ func (node *Node) tryBroadcastStaking(stakingTx *staking.StakingTransaction) {
...
-	// TODO: everyone should record staking txns, not just leader
-	if node.Consensus.IsLeader() &&
-		node.NodeConfig.ShardID == shard.BeaconChainShardID {
+func (node *Node) AddPendingStakingTransaction(newStakingTx *staking.StakingTransaction) {
+	if node.NodeConfig.ShardID == shard.BeaconChainShardID {
 		node.addPendingStakingTransactions(staking.StakingTransactions{newStakingTx})
-	} else {
-		utils.Logger().Info().Str("Hash", newStakingTx.Hash().Hex()).Msg("Broadcasting Staking Tx")
-		node.tryBroadcastStaking(newStakingTx)
+		if node.NodeConfig.Role() != nodeconfig.ExplorerNode {
```
Comment (rlan35) [inline node/node.go]: The original logic is still correct, we want any node that's not the leader to broadcast the staking transaction (not just the explorer node, it's a single point of failure)

Comment (Daniel-VDM) [reply node/node.go]: discussed in person, will also change regular tx logic so that everyone broadcasts in this PR

File: core/tx_pool.go (hunk 6 above)
Comment (Daniel-VDM) [reply core/tx_pool.go]: refactored the staking tx check and tx pool now calls it

File: core/staking_verifier.go [hunk 14]
```diff
@@ -0,0 +1,265 @@
...
+	wrapper.Snapshot.Epoch = epoch
+	wrapper.Snapshot.NumBlocksSigned = zero
+	wrapper.Snapshot.NumBlocksToSign = zero
+	return wrapper.SanityCheck()
+}
+
+// VerifyEditValidatorStakingMsg verifies the edit validator message using
+// the stateDB and chainContext if they are provided.
+// It is not necessary to provide the stateDB and/or chainContext; validation that
+// does not using said components will still run.
```
Comment (rlan35) [inline core/staking_verifier.go]: "does not use". Please correct everywhere.

Comment (Daniel-VDM) [reply core/staking_verifier.go]: corrected.

File: core/state_transition.go [hunk 15]
```diff
@@ -438,167 +432,94 @@ func (st *StateTransition) applyEditValidatorTx(
...
-	}
 	wrapper := st.state.GetStakingInfo(delegate.ValidatorAddress)
 	if wrapper == nil {
 		return errValidatorNotExist
 	}
-
-	stateDB := st.state
-	delegatorExist := false
-	for i := range wrapper.Delegations {
-		delegation := &wrapper.Delegations[i]
```
Comment (rlan35) [inline core/state_transition.go]: Use this original way of referencing delegation. Your way gets a copy of the delegation, and it is a source of problems

File: core/state_transition.go [hunk 16]
```diff
@@ -314,54 +311,63 @@ func (st *StateTransition) StakingTransitionDb() (usedGas uint64, err error) {
 		if err = rlp.DecodeBytes(msg.Data(), stkMsg); err != nil {
 			return 0, err
 		}
-		utils.Logger().Info().Msgf("[DEBUG STAKING] staking type: %s, txn: %+v", msg.Type(), stkMsg)
+		utils.Logger().Info().Msgf("[DEBUG STAKING] staking type: %s, gas: %d, txn: %+v", msg.Type(), gas, stkMsg)
 		if msg.From() != stkMsg.ValidatorAddress {
 			return 0, errInvalidSigner
 		}
+		if err := VerifyCreateValidatorStakingMsg(st.state, st.evm.EpochNumber, stkMsg); err != nil {
```
Comment (rlan35) [inline core/state_transition.go]: Better put the Verify.... inside apply........, so the apply can never mistakenly commit an invalid message.

File: core/staking_verifier.go [hunk 17]
```diff
@@ -0,0 +1,265 @@
...
+	}
+	if !stateDB.IsValidator(msg.ValidatorAddress) {
+		return errValidatorNotExist
+	}
+	wrapper := stateDB.GetStakingInfo(msg.ValidatorAddress)
+	if wrapper == nil {
+		return errValidatorNotExist
+	}
+	delegatorExist := false
+	for _, delegation := range wrapper.Delegations {
```
Comment (rlan35) [inline core/staking_verifier.go]: same here for the reference.

File: core/staking_verifier.go [hunk 18]
```diff
@@ -0,0 +1,265 @@
...
+			return errors.Wrapf(
+				errInsufficientBalanceForStake,
+				"total-delegated %s own-current-balance %s amount-to-delegate %s",
+				totalInUndelegation.String(),
+				balance.String(),
+				msg.Amount.String(),
+			)
+		}
+	}
+	if !delegatorExist {
```
Comment (rlan35) [inline core/staking_verifier.go]: why not remove delegatorExist here, but in the apply....?

File: core/state_transition.go (hunk 15 above)
Comment (Daniel-VDM) [reply core/state_transition.go]: oh alright, changed back to original way

File: core/state_transition.go (hunk 16 above)
Comment (Daniel-VDM) [reply core/state_transition.go]: sure, moved into apply....

File: core/staking_verifier.go (hunk 18 above)
Comment (Daniel-VDM) [reply core/staking_verifier.go]: oh, this can be removed. Done in latest

File: core/staking_verifier.go [hunk 19]
```diff
@@ -0,0 +1,262 @@
...
+// VerifyDelegateStakingMsg verifies the delegate message using the stateDB if
+// it is provided. Otherwise, it will only check components that don't require it.
+//
+// Note that these checks never update the stateDB, it only reads from stateDB.
+func VerifyDelegateStakingMsg(stateDB vm.StateDB, msg *staking.Delegate) error {
+	if msg.Amount.Sign() == -1 {
+		return errNegativeAmount
+	}
+	if stateDB == nil {
+		return nil
```
Comment (rlan35) [inline core/staking_verifier.go]: If there is no state, then should return error. Otherwise, it can be a false negative.

File: core/state_transition.go [hunk 20]
```diff
@@ -373,19 +367,13 @@ func (st *StateTransition) StakingTransitionDb() (usedGas uint64, err error) {
...
-	if val := createValidator.ValidatorAddress; st.state.IsValidator(val) {
-		return errors.Wrapf(errValidatorExist, common2.MustAddressToBech32(val))
-	}
-
-	if !CanTransfer(st.state, createValidator.ValidatorAddress, createValidator.Amount) {
-		return errInsufficientBalanceForStake
+func (st *StateTransition) verifyAndApplyCreateValidatorTx(
+	createValidator *staking.CreateValidator, blockNum *big.Int,
+) error {
+	err := VerifyCreateValidatorStakingMsg(st.state, st.evm.EpochNumber, createValidator)
```
Comment (rlan35) [inline core/state_transition.go]: You can return the created validator from VerifyCreateValidatorStakingMsg (maybe rename to VerifyAndCreateValidatorStakingMsg), and directly commit with that validator. Then you don't need to redo the whole wrapper.validator = .... ; wrapper.snapshot = .... thing. They are duplicate code.

File: core/state_transition.go [hunk 21]
```diff
@@ -410,13 +398,13 @@ func (st *StateTransition) applyCreateValidatorTx(createValidator *staking.Creat
 	return nil
 }
 
-func (st *StateTransition) applyEditValidatorTx(
+func (st *StateTransition) verifyAndApplyEditValidatorTx(
 	editValidator *staking.EditValidator, blockNum *big.Int,
 ) error {
-	if !st.state.IsValidator(editValidator.ValidatorAddress) {
-		return errValidatorNotExist
+	err := VerifyEditValidatorStakingMsg(st.state, st.bc, editValidator)
```
Comment (rlan35) [inline core/state_transition.go]: Same here. Can be VerifyAndEdit...... and return the edited validator object, so you don't have to do it again after.

File: core/state_transition.go [hunk 22]
```diff
@@ -438,167 +426,98 @@ func (st *StateTransition) applyEditValidatorTx(
...
 
-func (st *StateTransition) applyDelegateTx(delegate *staking.Delegate) error {
-	if delegate.Amount.Sign() == -1 {
-		return errNegativeAmount
-	}
-
-	if !st.state.IsValidator(delegate.ValidatorAddress) {
-		return errValidatorNotExist
+func (st *StateTransition) verifyAndApplyDelegateTx(delegate *staking.Delegate) error {
+	err := VerifyDelegateStakingMsg(st.state, delegate)
```
Comment (rlan35) [inline core/state_transition.go]: similar. We basically don't want to duplicate the complicated validator object manipulation logic twice, it's error-prone. Just do it once and use the modified object here to commit.

File: core/state_transition.go [hunk 23]
```diff
@@ -438,167 +426,98 @@ func (st *StateTransition) applyEditValidatorTx(
...
+	if err == nil {
+		st.state.SubBalance(delegate.DelegatorAddress, delegate.Amount)
 	}
+	return err
+}
 
-	if !st.state.IsValidator(undelegate.ValidatorAddress) {
-		return errValidatorNotExist
+func (st *StateTransition) verifyAndApplyUndelegateTx(undelegate *staking.Undelegate) error {
+	err := VerifyUndelegateStakingMsg(st.state, st.evm.EpochNumber, undelegate)
```
Comment (rlan35) [inline core/state_transition.go]: Same.

File: core/state_transition.go [hunk 24]
```diff
@@ -438,167 +426,98 @@ func (st *StateTransition) applyEditValidatorTx(
...
-			if delegation.Reward.Cmp(big.NewInt(0)) > 0 {
-				totalRewards.Add(totalRewards, delegation.Reward)
-			}
-
-			delegation.Reward.SetUint64(0)
-		}
-
-		err = st.state.UpdateStakingInfo(wrapper.Validator.Address, wrapper)
-		if err != nil {
+	updatedValidatorWrappers, totalRewards, err := PreProcessCollectRewardsFromDelegation(st.state, delegations)
```
Comment (rlan35) [inline core/state_transition.go]: Yup. this is how it should work. Just do the modification once.

File: core/types/tx_pool.go [hunk 25]
```diff
@@ -0,0 +1,102 @@
...
+	Hash() common.Hash
+	Nonce() uint64
+	ChainID() *big.Int
+	ToShardID() uint32
```
Comment (rlan35) [inline core/types/tx_pool.go]: where is ShardID() getter?

Comment (rlan35) [reply core/types/tx_pool.go]: For staking txn, there is no shard for it. But it implicitly means shard 0.

File: core/staking_verifier.go (hunk 19 above)
Comment (Daniel-VDM) [reply core/staking_verifier.go]: okay will return an error

File: core/types/tx_pool.go (hunk 25 above)
Comment (Daniel-VDM) [reply core/types/tx_pool.go]: yes, though it was used in tx pool, but just check and it does not. Will remove this requirement of the interface and remove the dummy implementation of the method in `StakingTransaction`

File: core/state_transition.go (hunk 20 above)
Comment (Daniel-VDM) [reply core/state_transition.go]: alright, refactored the checks to return components that need to be updated in the state

File: core/state_transition.go [hunk 26]
```diff
@@ -373,232 +366,68 @@ func (st *StateTransition) StakingTransitionDb() (usedGas uint64, err error) {
...
-
-	if newRate.GT(wrapper.Validator.MaxRate) {
-		return errCommissionRateChangeTooHigh
-	}
-
-	if err := st.state.UpdateStakingInfo(wrapper.Address, wrapper); err != nil {
+	if err := st.state.UpdateStakingInfo(wrapper.Validator.Address, wrapper); err != nil {
 		return err
 	}
+	st.state.SubBalance(delegate.DelegatorAddress, delegate.Amount)
```
Comment (rlan35) [inline core/state_transition.go]: This shouldn't be reducing by the whole delegate.Amount. The logic is this: to get the token for delegation, we first go into the list of undelegated tokens and use them first. If undelegated token are all used for delegation, then we use the account balance for the rest. So here it should be whatever is left in the "delegateBalance" in the original logic.

File: core/staking_verifier.go [hunk 27]
```diff
@@ -0,0 +1,279 @@
...
+	}
+	return updatedValidatorWrappers, totalRewards, nil
+}
+
+// getPendingEpoch returns the epoch of the next block given a chain
+func getPendingEpoch(chain interface{ CurrentBlock() *types.Block }) *big.Int {
+	currentBlock := chain.CurrentBlock()
+	pendingBlockNumber := new(big.Int).Add(currentBlock.Number(), big.NewInt(1))
+	blocksPerEpoch := new(big.Int).SetUint64(
+		shard.Schedule.InstanceForEpoch(currentBlock.Epoch()).BlocksPerEpoch(),
```
Comment (rlan35) [inline core/staking_verifier.go]: actually, don't add custom func like this. Just use the existing util "shard.Schedule.IsLastBlock". It will tell whether it's the last block of an epoch. From IsLastBlock, you can see the epoch calculation is not that straightforward due to some legacy nuances.

File: core/types/tx_pool.go (hunk 25 above)
Comment (rlan35) [reply core/types/tx_pool.go]: but don't you still need to check the shardID in txPool and remove those txns that's not for my shard?

File: core/state_transition.go (hunk 26 above)
Comment (Daniel-VDM) [reply core/state_transition.go]: Oh I see, okay will return the balance to be deducted from the delegator in `VerifyAndDelegateFromMsg`

File: core/staking_verifier.go (hunk 27 above)
Comment (Daniel-VDM) [reply core/staking_verifier.go]: oh alright, corrected for it.

File: core/types/tx_pool.go (hunk 25 above)
Comment (rlan35) [reply core/types/tx_pool.go]: How is this resolved? (Don't just resolve a comment without a clear closure).

Comment (Daniel-VDM) [reply core/types/tx_pool.go]: the current tx pool doesn't need shardIDs. Will implement this (and possibly Cx as well) in future PR.
